File: clients/go/internal/verify/shlex.go
package verify

import (
	"errors"
	"strings"
)

// posixSplit ports Python's shlex.split(s, posix=True) EXACTLY for the
// configuration opendaisugi.interpreter_parse actually uses:
// whitespace_split=True (set by shlex.split itself), commenters='' (comments
// argument defaults to False in shlex.split, so '#' is an ordinary
// character), punctuation_chars='' (shlex default).
//
// Under that configuration cpython's read_token state machine collapses:
// wordchars membership never matters (whitespace_split=True makes the
// wordchars branch and its fallback identical), so this is a small state
// machine over {space, word, single-quote, double-quote, escape} — see the
// derivation in the Go client's session notes / ADJUDICATIONS if this ever
// needs re-deriving. Returns an error exactly where Python raises ValueError
// ("No closing quotation" / "No escaped character") — the caller (
// parse_interpreter) treats that as "not parseable", matching the oracle's
// `except ValueError: return None`.
func posixSplit(s string) ([]string, error) {
	const (
		stSpace = iota
		stWord
		stSingle
		stDouble
		stEscape
	)
	runes := []rune(s)
	n := len(runes)
	i := 0
	state := stSpace
	escapedState := stWord
	var token []rune
	var tokens []string

	isWS := func(r rune) bool { return r == ' ' || r == '\t' || r == '\r' || r == '\n' }

	for {
		hasCh := i < n
		var ch rune
		if hasCh {
			ch = runes[i]
		}
		switch state {
		case stSpace:
			if !hasCh {
				return tokens, nil // clean EOF between tokens
			}
			i++
			switch {
			case isWS(ch):
				// stay in space state, nothing to do
			case ch == '\\':
				escapedState = stWord
				state = stEscape
			case ch == '\'':
				state = stSingle
			case ch == '"':
				state = stDouble
			default:
				token = append(token, ch)
				state = stWord
			}
		case stSingle:
			if !hasCh {
				return nil, errors.New("no closing quotation")
			}
			i++
			if ch == '\'' {
				state = stWord
			} else {
				token = append(token, ch)
			}
		case stDouble:
			if !hasCh {
				return nil, errors.New("no closing quotation")
			}
			i++
			if ch == '"' {
				state = stWord
			} else if ch == '\\' {
				escapedState = stDouble
				state = stEscape
			} else {
				token = append(token, ch)
			}
		case stEscape:
			if !hasCh {
				return nil, errors.New("no escaped character")
			}
			i++
			if escapedState == stDouble {
				// Inside double quotes, backslash is meaningful only before
				// '\\' or '"' — anything else keeps the backslash literally.
				if ch != '\\' && ch != '"' {
					token = append(token, '\\')
				}
				token = append(token, ch)
			} else {
				token = append(token, ch)
			}
			state = escapedState
		case stWord:
			if !hasCh {
				tokens = append(tokens, string(token))
				return tokens, nil
			}
			switch {
			case isWS(ch):
				i++
				tokens = append(tokens, string(token))
				token = nil
				state = stSpace
			case ch == '\'':
				i++
				state = stSingle
			case ch == '"':
				i++
				state = stDouble
			case ch == '\\':
				i++
				escapedState = stWord
				state = stEscape
			default:
				i++
				token = append(token, ch)
			}
		}
	}
}

// shlexQuote ports Python's shlex.quote exactly: empty string -> "''";
// a string with no characters outside [A-Za-z0-9_@%+=:,./-] is returned
// unquoted; otherwise it is single-quoted with embedded quotes escaped as
// '"'"'.
func shlexQuote(s string) string {
	if s == "" {
		return "''"
	}
	safe := true
	for _, r := range s {
		if !isShlexSafe(r) {
			safe = false
			break
		}
	}
	if safe {
		return s
	}
	return "'" + strings.ReplaceAll(s, "'", `'"'"'`) + "'"
}

func isShlexSafe(r rune) bool {
	switch {
	case r >= 'a' && r <= 'z', r >= 'A' && r <= 'Z', r >= '0' && r <= '9':
		return true
	}
	switch r {
	case '_', '@', '%', '+', '=', ':', ',', '.', '/', '-':
		return true
	}
	return false
}

// shlexJoin quotes and joins tokens with a single space — the Go analogue of
// interpreter_parse.py's `" ".join(shlex.quote(t) for t in tokens)`.
func shlexJoin(tokens []string) string {
	quoted := make([]string, len(tokens))
	for i, t := range tokens {
		quoted[i] = shlexQuote(t)
	}
	return strings.Join(quoted, " ")
}

File: clients/ts/src/shlex.ts
/**
 * Port of CPython 3.12 `shlex.split(s, comments=False, posix=True)` and
 * `shlex.quote`. Only the reachable configuration is ported (posix=True,
 * whitespace_split=True, commenters='', no punctuation_chars) — the general
 * shlex class has many modes the oracle never exercises.
 *
 * `shlexSplitPosix` throws `ShlexError` on unbalanced quotes / a trailing
 * unescaped backslash, matching CPython's `ValueError` — callers that treat
 * a parse failure as "not an interpreter" should catch it (see
 * `interpreterParse.ts`, mirroring `parse_interpreter`'s `except ValueError`).
 */

export class ShlexError extends Error {}

const WHITESPACE = new Set([" ", "\t", "\r", "\n"]);
const QUOTES = new Set(["'", '"']);
const ESCAPE = "\\";
const ESCAPED_QUOTES = new Set(['"']); // only double-quotes honor backslash-escaping

type State = " " | "a" | "'" | '"' | "\\" | null;

export function shlexSplitPosix(s: string): string[] {
  const tokens: string[] = [];
  let i = 0;
  const n = s.length;

  while (true) {
    let state: State = " ";
    let token = "";
    let quoted = false;
    let escapedState: State = " ";
    let sawEnd = false;

    while (true) {
      const nextchar = i < n ? s[i]! : "";
      if (nextchar !== "") i += 1;

      if (state === null) {
        token = "";
        break;
      } else if (state === " ") {
        if (nextchar === "") {
          state = null;
          break;
        } else if (WHITESPACE.has(nextchar)) {
          if (token || quoted) {
            break;
          } else {
            continue;
          }
        } else if (nextchar === ESCAPE) {
          escapedState = "a";
          state = "\\";
        } else if (QUOTES.has(nextchar)) {
          state = nextchar as State;
        } else {
          // whitespace_split catch-all (wordchars are a strict subset of
          // this in the posix+whitespace_split configuration, so folding
          // them together is behavior-preserving).
          token = nextchar;
          state = "a";
        }
      } else if (state === "'" || state === '"') {
        quoted = true;
        if (nextchar === "") {
          throw new ShlexError("No closing quotation");
        }
        if (nextchar === state) {
          state = "a";
        } else if (nextchar === ESCAPE && ESCAPED_QUOTES.has(state)) {
          escapedState = state;
          state = "\\";
        } else {
          token += nextchar;
        }
      } else if (state === "\\") {
        if (nextchar === "") {
          throw new ShlexError("No escaped character");
        }
        if (
          escapedState !== null &&
          QUOTES.has(escapedState) &&
          nextchar !== state &&
          nextchar !== escapedState
        ) {
          token += state; // preserve the backslash literally
        }
        token += nextchar;
        state = escapedState;
      } else if (state === "a") {
        if (nextchar === "") {
          state = null;
          break;
        } else if (WHITESPACE.has(nextchar)) {
          state = " ";
          if (token || quoted) {
            break;
          } else {
            continue;
          }
        } else if (QUOTES.has(nextchar)) {
          state = nextchar as State;
        } else if (nextchar === ESCAPE) {
          escapedState = "a";
          state = "\\";
        } else {
          token += nextchar;
        }
      }
    }

    let result: string | null = token;
    if (!quoted && result === "") {
      result = null;
    }
    if (result === null) {
      sawEnd = true;
    } else {
      tokens.push(result);
    }
    if (sawEnd) break;
  }

  return tokens;
}

// shlex.quote — mirrors CPython's `_find_unsafe` (ASCII: not in [\w@%+=:,./-]).
const UNSAFE_RE = /[^A-Za-z0-9_@%+=:,./-]/;

export function shlexQuote(s: string): string {
  if (s === "") return "''";
  if (!UNSAFE_RE.test(s)) return s;
  return "'" + s.split("'").join("'\"'\"'") + "'";
}
